File: pzb_ros/src/pzb_traffic/scripts/color_detector_node.py
#!/usr/bin/env python3
"""
Color detector node for traffic light detection.

Subscribes:
  /camera/image_compressed  (sensor_msgs/CompressedImage)

Publishes:
  /traffic_light_color         (std_msgs/String)  "red" | "green" | "yellow" | "none"
  /traffic_detector/debug_image (sensor_msgs/Image)  annotated frame
"""

import signal
import numpy as np
import cv2

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy, HistoryPolicy

from sensor_msgs.msg import CompressedImage, Image
from std_msgs.msg import String

_BEST_EFFORT_QOS = QoSProfile(
    reliability=ReliabilityPolicy.BEST_EFFORT,
    durability=DurabilityPolicy.VOLATILE,
    history=HistoryPolicy.KEEP_LAST,
    depth=1,
)

# BGR colors for debug annotation
_DEBUG_COLORS = {
    'red':    (0,   0,   220),
    'green':  (0,   200, 0),
    'yellow': (0,   200, 200),
    'none':   (150, 150, 150),
}


class ColorDetectorNode(Node):

    def __init__(self):
        super().__init__('color_detector_node')

        # HSV range params
        self.declare_parameter('red_h_low1',  0)
        self.declare_parameter('red_h_high1', 10)
        self.declare_parameter('red_h_low2',  165)
        self.declare_parameter('red_h_high2', 180)
        self.declare_parameter('red_s_min',   100)
        self.declare_parameter('red_v_min',   80)

        self.declare_parameter('green_h_low',  45)
        self.declare_parameter('green_h_high', 85)
        self.declare_parameter('green_s_min',  80)
        self.declare_parameter('green_v_min',  80)

        self.declare_parameter('yellow_h_low',  18)
        self.declare_parameter('yellow_h_high', 35)
        self.declare_parameter('yellow_s_min',  120)
        self.declare_parameter('yellow_v_min',  100)

        self.declare_parameter('min_blob_area',  800.0)
        self.declare_parameter('confirm_frames', 3)
        self.declare_parameter('loop_hz',        30.0)

        self._min_area    = float(self.get_parameter('min_blob_area').value)
        self._confirm_n   = int(self.get_parameter('confirm_frames').value)

        # Confirmation filter state
        self._candidate       = 'none'   # color being accumulated
        self._candidate_count = 0        # consecutive frames of candidate
        self._confirmed_color = 'none'   # last published color

        self._morph_kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (5, 5))
        self._clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))

        self.create_subscription(
            CompressedImage,
            '/camera/image_compressed',
            self._image_callback,
            _BEST_EFFORT_QOS,
        )

        self._color_pub = self.create_publisher(String, '/traffic_light_color', 10)
        self._debug_pub = self.create_publisher(Image, '/traffic_detector/debug_image', 1)

        self.get_logger().info(
            f'ColorDetectorNode ready — min_blob_area={self._min_area}, '
            f'confirm_frames={self._confirm_n}'
        )

    # ------------------------------------------------------------------ helpers

    def _hsv_ranges(self):
        """Return per-color list of (lower, upper) HSV numpy arrays."""
        g = self.get_parameter
        red = [
            (np.array([g('red_h_low1').value,  g('red_s_min').value, g('red_v_min').value]),
             np.array([g('red_h_high1').value, 255, 255])),
            (np.array([g('red_h_low2').value,  g('red_s_min').value, g('red_v_min').value]),
             np.array([g('red_h_high2').value, 255, 255])),
        ]
        green = [
            (np.array([g('green_h_low').value,  g('green_s_min').value, g('green_v_min').value]),
             np.array([g('green_h_high').value, 255, 255])),
        ]
        yellow = [
            (np.array([g('yellow_h_low').value,  g('yellow_s_min').value, g('yellow_v_min').value]),
             np.array([g('yellow_h_high').value, 255, 255])),
        ]
        return {'red': red, 'green': green, 'yellow': yellow}

    def _largest_blob_area(self, mask) -> float:
        """Return the area of the largest contour in the binary mask."""
        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        if not contours:
            return 0.0
        return float(max(cv2.contourArea(c) for c in contours))

    def _build_mask(self, hsv, ranges):
        """OR together all (lower, upper) range masks and apply morphological open."""
        combined = None
        for lower, upper in ranges:
            m = cv2.inRange(hsv, lower, upper)
            combined = m if combined is None else cv2.bitwise_or(combined, m)
        # Morphological open: removes isolated noise pixels
        return cv2.morphologyEx(combined, cv2.MORPH_OPEN, self._morph_kernel, iterations=2)

    def _draw_debug(self, frame, detected_color, areas):
        """Annotate frame with detected color and blob areas."""
        bgr = _DEBUG_COLORS.get(detected_color, (150, 150, 150))
        cv2.putText(frame, f'Color: {detected_color.upper()}',
                    (10, 35), cv2.FONT_HERSHEY_SIMPLEX, 1.0, bgr, 2)
        y = 65
        for name, area in areas.items():
            cv2.putText(frame, f'{name}: {area:.0f}px2',
                        (10, y), cv2.FONT_HERSHEY_SIMPLEX, 0.55,
                        _DEBUG_COLORS[name], 1)
            y += 22
        return frame

    # ------------------------------------------------------------------ callback

    def _image_callback(self, msg: CompressedImage):
        buf = np.frombuffer(msg.data, np.uint8)
        frame = cv2.imdecode(buf, cv2.IMREAD_COLOR)
        if frame is None:
            return

        # Bilateral filter: reduces noise while preserving color edges (better than Gaussian for screen glare)
        blurred = cv2.bilateralFilter(frame, d=9, sigmaColor=75, sigmaSpace=75)
        hsv = cv2.cvtColor(blurred, cv2.COLOR_BGR2HSV)
        # CLAHE on V channel normalizes brightness variation from screen angle/glare
        h, s, v = cv2.split(hsv)
        v = self._clahe.apply(v)
        hsv = cv2.merge([h, s, v])

        ranges = self._hsv_ranges()
        areas = {}
        for color_name, color_ranges in ranges.items():
            mask = self._build_mask(hsv, color_ranges)
            areas[color_name] = self._largest_blob_area(mask)

        # Winning color: largest area above threshold
        best_color = 'none'
        best_area = self._min_area
        for color_name, area in areas.items():
            if area > best_area:
                best_area = area
                best_color = color_name

        # Confirmation filter: require N consecutive frames
        if best_color == self._candidate:
            self._candidate_count += 1
        else:
            self._candidate = best_color
            self._candidate_count = 1

        if self._candidate_count >= self._confirm_n:
            self._confirmed_color = self._candidate

        # Publish confirmed color
        color_msg = String()
        color_msg.data = self._confirmed_color
        self._color_pub.publish(color_msg)

        # Publish debug image if anyone is listening
        if self._debug_pub.get_subscription_count() > 0:
            annotated = self._draw_debug(frame.copy(), self._confirmed_color, areas)
            debug_msg = Image()
            debug_msg.header = msg.header
            debug_msg.height = annotated.shape[0]
            debug_msg.width  = annotated.shape[1]
            debug_msg.encoding = 'bgr8'
            debug_msg.step = annotated.shape[1] * 3
            debug_msg.data = annotated.tobytes()
            self._debug_pub.publish(debug_msg)


def main(args=None):
    rclpy.init(args=args)
    node = ColorDetectorNode()

    def _shutdown(signum, frame):
        rclpy.try_shutdown()

    signal.signal(signal.SIGTERM, _shutdown)

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.try_shutdown()


if __name__ == '__main__':
    main()

File: pzb_ros/src/pzb_traffic_cpp/src/color_detector_node.cpp
// C++ port of color_detector_node.py — HSV traffic-light detector. Drop-in: same node
// name (color_detector_node), same params (reuses traffic_params.yaml), same topics
// (/camera/image_raw -> /traffic_light_color). FAITHFUL: crop top 2/3 -> resize (320,160)
// -> GaussianBlur -> BGR2HSV -> CLAHE on V -> per-color inRange OR-masks -> MORPH_OPEN x2
// -> largest round/bright/aspect-valid blob -> winner above min_area -> N-frame confirm.
// Every-3rd-frame skip preserved.
#include <map>
#include <string>
#include <vector>

#include <cv_bridge/cv_bridge.h>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <std_msgs/msg/string.hpp>

class ColorDetectorNode : public rclcpp::Node {
 public:
  ColorDetectorNode() : rclcpp::Node("color_detector_node") {
    auto gi = [&](const std::string& n, int d) { return this->declare_parameter(n, d); };
    red_h_low1_ = gi("red_h_low1", 0);   red_h_high1_ = gi("red_h_high1", 10);
    red_h_low2_ = gi("red_h_low2", 165); red_h_high2_ = gi("red_h_high2", 180);
    red_s_min_  = gi("red_s_min", 100);  red_v_min_   = gi("red_v_min", 80);
    green_h_low_ = gi("green_h_low", 45); green_h_high_ = gi("green_h_high", 85);
    green_s_min_ = gi("green_s_min", 80); green_v_min_  = gi("green_v_min", 80);
    yellow_h_low_ = gi("yellow_h_low", 18); yellow_h_high_ = gi("yellow_h_high", 35);
    yellow_s_min_ = gi("yellow_s_min", 120); yellow_v_min_ = gi("yellow_v_min", 100);
    min_area_  = this->declare_parameter("min_blob_area", 800.0);
    confirm_n_ = this->declare_parameter("confirm_frames", 3);
    std::string in_topic = this->declare_parameter("input_topic", std::string("/camera/image_raw"));

    morph_kernel_ = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(5, 5));
    clahe_ = cv::createCLAHE(2.0, cv::Size(8, 8));

    auto best_effort = rclcpp::QoS(rclcpp::KeepLast(1)).best_effort();
    sub_ = create_subscription<sensor_msgs::msg::Image>(
        in_topic, best_effort, [this](sensor_msgs::msg::Image::SharedPtr m) { on_image(m); });
    pub_ = create_publisher<std_msgs::msg::String>("/traffic_light_color", 10);
    RCLCPP_INFO(get_logger(), "C++ color_detector_node ready (min_area=%.0f, confirm=%d)",
                min_area_, confirm_n_);
  }

 private:
  // OR all (lower,upper) HSV ranges then MORPH_OPEN x2 (matches _build_mask).
  cv::Mat build_mask(const cv::Mat& hsv,
                     const std::vector<std::pair<cv::Scalar, cv::Scalar>>& ranges) {
    cv::Mat combined;
    for (auto& [lo, hi] : ranges) {
      cv::Mat m; cv::inRange(hsv, lo, hi, m);
      if (combined.empty()) combined = m; else cv::bitwise_or(combined, m, combined);
    }
    cv::Mat opened;
    cv::morphologyEx(combined, opened, cv::MORPH_OPEN, morph_kernel_, cv::Point(-1, -1), 2);
    return opened;
  }

  // Largest round/bright-aspect blob area above min_area (matches _largest_blob_area:
  // area>=min, circularity>=0.35, aspect in (0.5,2.0)).
  double largest_blob_area(const cv::Mat& mask) {
    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
    double best = 0.0;
    for (auto& c : contours) {
      double area = cv::contourArea(c);
      if (area < min_area_) continue;
      double perim = cv::arcLength(c, true);
      if (perim == 0) continue;
      double circ = 4.0 * M_PI * area / (perim * perim);
      if (circ < 0.35) continue;
      cv::Rect r = cv::boundingRect(c);
      double aspect = r.height > 0 ? static_cast<double>(r.width) / r.height : 0.0;
      if (!(aspect > 0.5 && aspect < 2.0)) continue;
      if (area > best) best = area;
    }
    return best;
  }

  void on_image(sensor_msgs::msg::Image::SharedPtr msg) {
    frame_skip_ = (frame_skip_ + 1) % 3;
    if (frame_skip_ != 0) return;

    cv::Mat full;
    try { full = cv_bridge::toCvShare(msg, "bgr8")->image; }
    catch (const std::exception& e) { RCLCPP_WARN(get_logger(), "cv_bridge: %s", e.what()); return; }

    cv::Mat top = full(cv::Range(0, full.rows * 2 / 3), cv::Range::all());
    cv::Mat frame; cv::resize(top, frame, cv::Size(320, 160), 0, 0, cv::INTER_AREA);
    cv::Mat blurred; cv::GaussianBlur(frame, blurred, cv::Size(5, 5), 0);
    cv::Mat hsv; cv::cvtColor(blurred, hsv, cv::COLOR_BGR2HSV);
    std::vector<cv::Mat> ch; cv::split(hsv, ch);
    clahe_->apply(ch[2], ch[2]);
    cv::merge(ch, hsv);

    std::map<std::string, std::vector<std::pair<cv::Scalar, cv::Scalar>>> ranges{
      {"red", {{cv::Scalar(red_h_low1_, red_s_min_, red_v_min_), cv::Scalar(red_h_high1_, 255, 255)},
               {cv::Scalar(red_h_low2_, red_s_min_, red_v_min_), cv::Scalar(red_h_high2_, 255, 255)}}},
      {"green", {{cv::Scalar(green_h_low_, green_s_min_, green_v_min_), cv::Scalar(green_h_high_, 255, 255)}}},
      {"yellow", {{cv::Scalar(yellow_h_low_, yellow_s_min_, yellow_v_min_), cv::Scalar(yellow_h_high_, 255, 255)}}},
    };

    std::map<std::string, double> areas;
    for (auto& [name, r] : ranges) areas[name] = largest_blob_area(build_mask(hsv, r));

    std::string best_color = "none";
    double best_area = min_area_;
    // iterate in a fixed order matching Python dict insertion (red, green, yellow)
    for (const std::string& name : {"red", "green", "yellow"})
      if (areas[name] > best_area) { best_area = areas[name]; best_color = name; }

    if (best_color == candidate_) candidate_count_++;
    else { candidate_ = best_color; candidate_count_ = 1; }
    if (candidate_count_ >= confirm_n_) confirmed_color_ = candidate_;

    std_msgs::msg::String m; m.data = confirmed_color_; pub_->publish(m);
  }

  int red_h_low1_, red_h_high1_, red_h_low2_, red_h_high2_, red_s_min_, red_v_min_;
  int green_h_low_, green_h_high_, green_s_min_, green_v_min_;
  int yellow_h_low_, yellow_h_high_, yellow_s_min_, yellow_v_min_;
  double min_area_; int confirm_n_;
  int frame_skip_ = 0, candidate_count_ = 0;
  std::string candidate_ = "none", confirmed_color_ = "none";
  cv::Mat morph_kernel_;
  cv::Ptr<cv::CLAHE> clahe_;
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ColorDetectorNode>());
  rclcpp::shutdown();
  return 0;
}
